rllib/utils/exploration/parameter_noise.py

```diff
@@ -0,0 +1,386 @@
+
+        inputs, dist_class, _ = policy.compute_distribution_inputs(
+            obs_batch=sample_batch[SampleBatch.CUR_OBS],
+            # TODO(sven): What about state-ins and seq-lens?
+            prev_action_batch=sample_batch.get(SampleBatch.PREV_ACTIONS),
+            prev_reward_batch=sample_batch.get(SampleBatch.PREV_REWARDS),
+            explore=not self.weights_are_currently_noisy)
+
+        # Categorical case (e.g. DQN).
+        if dist_class is Categorical:
+            action_dist = softmax(inputs)
+
+        if not self.weights_are_currently_noisy:
+            noisy_action_dist = action_dist
+        else:
+            noise_free_action_dist = action_dist
+
+        # Categorical case (e.g. DQN).
+        if dist_class is Categorical:
+            # Calculate KL-divergence (DKL(clean||noisy)) according to [2].
+            kl_divergence = np.nanmean(
+                np.sum(
+                    noise_free_action_dist *
+                    np.log(noise_free_action_dist /
+                           (noisy_action_dist + SMALL_NUMBER)), 1))
+            current_epsilon = self.sub_exploration.get_info()["cur_epsilon"]
+            if tf_sess is not None:
+                current_epsilon = tf_sess.run(current_epsilon)
+            delta = -np.log(1 - current_epsilon +
+                            current_epsilon / self.action_space.n)
+            if kl_divergence <= delta:
+                self.stddev_val *= 1.01
+            else:
+                self.stddev_val /= 1.01
+
+        # Set self.stddev to calculated value.
+        if self.framework == "tf":
+            self.stddev.load(self.stddev_val, session=tf_sess)
+        else:
+            self.stddev = self.stddev_val
+
+        return sample_batch
+
+    def _sample_new_noise(self, *, tf_sess=None):
+        """Samples new noise and stores it in `self.noise`."""
+        if self.framework == "tf":
+            if tf.executing_eagerly():
+                self._tf_sample_new_noise_op()
+            else:
+                tf_sess.run(self.tf_sample_new_noise_op)
+        else:
+            for i in range(len(self.noise)):
+                self.noise[i] = torch.normal(
+                    0.0, self.stddev, size=self.noise[i].size)
+
+    def _tf_sample_new_noise_op(self):
+        added_noises = []
+        for noise in self.noise:
+            added_noises.append(
+                tf.assign(
+                    noise,
+                    tf.random_normal(
+                        shape=noise.shape,
+                        stddev=self.stddev,
+                        dtype=tf.float32)))
+        return tf.group(*added_noises)
+
+    def _sample_new_noise_and_add(self, *, tf_sess=None, override=False):
+        if self.framework == "tf" and not tf.executing_eagerly():
+            if override and self.weights_are_currently_noisy:
+                tf_sess.run(self.tf_remove_noise_op)
+            tf_sess.run(self.tf_sample_new_noise_and_add_op)
```

Comment:
Is it possible to use the `utils/make_tf_callable` decorator to avoid having to call different code for eager vs graph?

Reply:
I'll check...

Reply:
I would need the session already in the `make_tf_callable` decorator.

Reply:
I think you can get that with `get_default_session()` in the exploration constructor, which should always be set correctly when the policy is being constructed.

Reply:
I don't think this will work. What if I need to use with `tf.control_dependencies()` in a  `make_tf_callable` func? Eager would complain, I think.

Reply:
Let me try ...